instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ContainerResource.java

```diff
@@ -77,19 +77,83 @@ public static Resource get() {
   // Exists for testing
   static class Filesystem {
 
+    private static final Logger logger = Logger.getLogger(Filesystem.class.getName());
+
     boolean isReadable(Path path) {
       return Files.isReadable(path);
     }
 
     @MustBeClosed
     Stream<String> lines(Path path) throws IOException {
-      return Files.lines(path);
+
+      String osName = getOsName();
+      if (!(osName.equalsIgnoreCase("z/OS") || osName.equalsIgnoreCase("OS/390"))) {
+        return Files.lines(path);
+      } else {
+        return zosLines(path);
+      }
+    }
+
+    // This method reads the /proc system using the correct Charset for z/OS
+    @MustBeClosed
+    Stream<String> zosLines(Path path) throws IOException {
+
+      List<Charset> charsetsToTest = new ArrayList<Charset>();
+
+      // Since this class needs to be compatible with Java9&11, use the 'native.encoding' property
+      // to get the nativeCharset on java17+
+      // rather than System.console().charset().
+      if (System.getProperty("native.encoding") != null) {
+        Charset nativeCharset = Charset.forName(System.getProperty("native.encoding"));
+        if (nativeCharset != null) {
+          charsetsToTest.add(nativeCharset);
+        }
+      } else {
+        // On older javas that property won't exist but the default charset will be the native
+        // encoding.
+        charsetsToTest.add(Charset.defaultCharset());
+      }
+
+      // As a fallback value, add a hardcoded reference to IBM1047 (Canonical name: Cp1047).
+      Charset ibm1047 = Charset.forName("Cp1047");
+      if (ibm1047 != null) { // safety check
+        charsetsToTest.add(ibm1047);
+      }
+
+      // The odds of /proc being UTF-8 but the native encoding saying otherwise is extremely low
+      // but since UTF-8 is the standard it should be checked.
+      charsetsToTest.add(StandardCharsets.UTF_8);
```

Comment:
does z/os ever use utf-8 for this file? assuming it does, when you'd attempt to parse utf-8 encoded file with `Cp1047` would that ever fail? is it possible for the code flow to reach a point where it attempts utf-8 at all?

Reply:
It cannot, and I did ask if there was any chance of it happening in the future and that's unlikely too. But I put this line in there anyway because I felt that the industry standard charset should always be on the list. If you disagree say the word and I'll remove it.

Reply:
I think you should remove it.